src/Features/CSharp/Portable/ConvertLinq/CSharpConvertLinqQueryToForEachProvider.cs

```diff
@@ -0,0 +1,645 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using Microsoft.CodeAnalysis.ConvertLinq;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.Simplification;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertLinq
+{
+    internal sealed class CSharpConvertLinqQueryToLinqMethodProvider : AbstractConvertLinqProvider
+    {
+        private static readonly TypeSyntax VarNameIdentifier = SyntaxFactory.IdentifierName("var");
+
+        protected override IAnalyzer CreateAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+            => new CSharpAnalyzer(semanticModel, cancellationToken);
+
+        private sealed class CSharpAnalyzer : AnalyzerBase<QueryExpressionSyntax, SyntaxNode, SyntaxNode>
+        {
+            public CSharpAnalyzer(SemanticModel semanticModel, CancellationToken cancellationToken)
+                : base(semanticModel, cancellationToken)
+            {
+            }
+
+            protected override string Title => CSharpFeaturesResources.Convert_linq_query_to_foreach;
+
+            protected override bool TryConvert(QueryExpressionSyntax source, out DocumentUpdate documentUpdate)
+            {
+                if (_semanticModel.GetDiagnostics(source.Span, _cancellationToken).Any())
+                {
+                    documentUpdate = default;
+                    return false;
+                }
+
+                if (source.DescendantTrivia().Any(trivia => trivia.MatchesKind(
+                    SyntaxKind.SingleLineCommentTrivia,
+                    SyntaxKind.MultiLineCommentTrivia,
+                    SyntaxKind.MultiLineDocumentationCommentTrivia) ||
```

Comment:
Let us consider them separately. Talked to @heejaechang@outlook.com who mentioned it could be a non-trivial task to migrate comments from one code to another.

[https://github.com/dotnet/roslyn/issues/25639](https://github.com/dotnet/roslyn/issues/25639)

---
In reply to: [175267407](https://github.com/dotnet/roslyn/pull/25362#discussion_r175267407) [](ancestors = 175267407)

Reply:
I'm fine with us creating tracking work to do this better... 

However, my overall point is that until that work is done we should default to *supporting* the feature, not disabling it.

Reply:
I'm not sure that we support comments automatically with this conversion. To avoid breaking user (= losing comments), likely we should disable this by default.

---
In reply to: [175950029](https://github.com/dotnet/roslyn/pull/25362#discussion_r175950029) [](ancestors = 175950029)

Reply:
> likely we should disable this by default.

Sorry i wasn't clear.  By default, the IDE defaults to offering tehse sorts of features, even if they have comments.  That's because we don't want users to not understand why they're not showing up.  We do this even when features have comments and even when we do not know how to carry those comments over.

This is because refactoring is just a helper tool.  The user wants to go to a location and make it happen, often even when it is not perfect.  They can manually tweak things afterwards if necessary, or just not apply the refactoring if they don't like it.

However, if the feature doesn't even appear tehy just feel like the product is broken as there is no reason for them to think "oh, because i had a comment, that will make things unavailable".

>  To avoid breaking user (= losing comments)

Note: we do disable when code would truly 'break'.  i..e if there's a pp-directive interspersed with the code that we don't know what to do with.  However, not perfectly carrying a comment over is not considered a 'break', and we will give users the refactoring, allowing them to decide if the change is acceptable or not.  

--

In other words: provide the help, allow the users to decide if they're ok with the change.  Making the decision for them just makes the product seem flakey in ways that they cannot diagnose.  "it works for me in this large query over here.... but not over here... well, guess this is just a buggy feature."

Note: if you are really worried about this, you can add "Warning annotations" on the comments, calling the user's attention to the fact that their comments might go missing.  This is what i did with "convert for to foreach" when we allowed the user to run the feature even in circumstances that might potentially have deeper impact than they might realize.

Reply:
Thank you! There will be a design discussion soon to agree a consistent approach to this. I'll address this in the next PR after the discussion complete.

---
In reply to: [176271009](https://github.com/dotnet/roslyn/pull/25362#discussion_r176271009) [](ancestors = 176271009)